pkg/build/controller/config_controller.go

```diff
@@ -6,19 +6,38 @@ import (
 	"github.com/golang/glog"
 	kapi "k8s.io/kubernetes/pkg/api"
 	kerrors "k8s.io/kubernetes/pkg/api/errors"
-	"k8s.io/kubernetes/pkg/util"
+	kutil "k8s.io/kubernetes/pkg/util"
 
 	buildapi "github.com/openshift/origin/pkg/build/api"
 	buildclient "github.com/openshift/origin/pkg/build/client"
+	"github.com/openshift/origin/pkg/build/util"
 )
 
 type BuildConfigController struct {
 	BuildConfigInstantiator buildclient.BuildConfigInstantiator
+	BuildConfigDeleter      buildclient.BuildConfigDeleter
+	BuildDeleter            buildclient.BuildDeleter
+	BuildLister             buildclient.BuildLister
 }
 
 func (c *BuildConfigController) HandleBuildConfig(bc *buildapi.BuildConfig) error {
 	glog.V(4).Infof("Handling BuildConfig %s/%s", bc.Namespace, bc.Name)
 
+	if !bc.DeletionTimestamp.IsZero() {
```

Comment:
don't you need to return a non-zero value from CheckGracefulDelete?

but i think gracefuldelete will be a problematic way to solve this since we do not want to delete the buildconfig as long as there are any builds that weren't deleted yet.  so you don't want the grace period to expire and k8s to delete the buildconfig while you're still trying to delete builds, or you'll never delete the builds.

I don't know that you need an additional phase, but i think you do need to mark the buildconfig as deletepending (possibly with the timestamp) and have this controller actually delete the buildconfing once all the builds are successfully deleted.

you'll need to make sure HandleBuildConfig doesn't get called on the delete event, or course, or otherwise do something to ensure you don't create a loop.